src/ray/common/ray_syncer/ray_syncer-inl.h

```diff
@@ -136,58 +143,123 @@ class NodeSyncConnection {
   /// We'll filter the received or sent messages when the message is stale.
   absl::flat_hash_map<std::string, std::array<int64_t, kComponentArraySize>>
       node_versions_;
+
+  bool sending_ = false;
 };
 
-/// SyncConnection for gRPC server side. It has customized logic for sending.
-class ServerSyncConnection : public NodeSyncConnection {
+template <typename T>
+class BidiReactor : public T, public NodeSyncConnection {
```

Comment:
does it make sense BidiReactor to use composition (has the connection as part of its member) instead of inheritance? it's a little bit hard to follow the logic between reactor and connection.

Reply:
@scv119 It's according to the gRPC API :( 

Every RPC call is a stream, and unary is just a send and a response.

Here it's a Bidi, basically, multiple sends and multiple responses.

Connection here means logical connection, not physical one.

Maybe I should rename NodeSyncConnection to NodeSyncStreamBase, how do you think?